api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java

```diff
@@ -0,0 +1,309 @@
+				conditionConcept.setNonCoded(nonCodedConcept);
+			}
+			condition.setCondition(conditionConcept);
+			condition.setClinicalStatus(getStatus(context, submission));
+			condition.setOnsetDate(onSetDate.getValue(context, submission));
+			if (ConditionClinicalStatus.INACTIVE == getStatus(context, submission)) {
+				condition.setEndDate(endDate.getValue(context, submission));
+			}
+			condition.setPatient(session.getPatient());
+			conditionService = Context.getConditionService();
+			conditionService.saveCondition(condition);
+		}
+	}
+
+	@Override
+	public Collection<FormSubmissionError> validateSubmission(FormEntryContext context, HttpServletRequest submission) {
+		
+		// TODO Localize error messages
+		
+		List<FormSubmissionError> ret = new ArrayList<FormSubmissionError>();
+		Date givenOnsetDate = onSetDate.getValue(context, submission);
+		String condition = submission.getParameter(context.getFieldName(conditionName));
+		
+		if (context.getMode() != Mode.VIEW) {
+			
+			if (StringUtils.isBlank(condition)) {
+				ret.add(new FormSubmissionError(context
+	                    .getFieldName(conditionName), "Condition name required"));
+			}
+			
+			if (givenOnsetDate != null) {
+				if (givenOnsetDate.after(new Date())) {
+					ret.add(new FormSubmissionError(context
+	                        .getFieldName(onSetDate), "Onset date can't be after Today"));
+				}
+			} else {
+				ret.add(new FormSubmissionError(context
+	                    .getFieldName(onSetDate), "Onset date required"));
+			}
+			
+			ConditionClinicalStatus status = getStatus(context, submission);
+			if (status != null) {
+				if (ConditionClinicalStatus.INACTIVE == status) {
+					Date givenEndDate = endDate.getValue(context, submission);
+					if (givenEndDate != null) {
+						if (givenOnsetDate != null) {
+							if (givenOnsetDate.after(givenEndDate)) {
+								ret.add(new FormSubmissionError(context
+				                        .getFieldName(endDate), "End date can't be before onset date"));
+							}
+
+						} 
+					} else {
+						ret.add(new FormSubmissionError(context
+		                        .getFieldName(endDate), "End date required"));
+					}
+				}	
+		    } else {
+		    	ret.add(new FormSubmissionError(context
+	                    .getFieldName(conditionStatusWidget), "A Condition status is required"));
+		    }
+		}
+		return ret;
+	}
+
+	@Override
+	public String generateHtml(FormEntryContext context) {
+		mss = Context.getMessageSourceService();		
+		StringBuilder ret = new StringBuilder();
+		ret.append("<h5>");
+		ret.append(mss.getMessage("coreapps.conditionui.addNewCondition"));
+		ret.append("</h5>");
+		ret.append(createConditionNameWidget(context));
+		ret.append(createConditionDateWidgets(context));
+		ret.append(createConditionStatusWidgets(context));		
+		return ret.toString();
+	}
+		
+	public String createConditionNameWidget(FormEntryContext context) {
```

Comment:
I see that you changed this signature, but why has this become `public` then?

I would rather see a `protected` method named `htmlForConditionSearchWidget(FormEntryContext)` so that it reads like this:
```java
ret.append(htmlForConditionSearchWidget(context));
```

Reply:
Ok man, that naming sounds just perfect.

>I see that you changed this signature, but why has this become public then?

Actually that was a mistake, your right anyway